File: stacks/linked_list_stack.cpp
#include <iostream>

using namespace std;

struct node {
  int data;
  node* next;
  
  node(int data, node* next) {
    this->data = data;
    this->next = next;
  }
};

class stack {
  private:
    node* top_node;

  public:
    stack();
    void push(int);
    int pop();
    int peek();
    bool is_empty();
};

stack::stack() {
  top_node = NULL;
}

void stack::push(int x) {
  top_node = new node(x, top_node);
}

int stack::pop() {
  if(is_empty()) cout << "Stack underflow!" << endl;
  else {
    int data = top_node->data;
    top_node = top_node->next;
    return data;
  }
}

int stack::peek() {
  if(is_empty()) {
    cout << "Stack is empty!" << endl;
    return 0;
  }
  else return top_node->data;
}

bool stack::is_empty() {
  if(top_node == NULL) return true;
  else return false;
}

// test
int main() {
  stack my_stack;
  my_stack.push(5);
  my_stack.push(-7);
  cout << my_stack.pop() << endl;
  cout << my_stack.peek() << endl;
  cout << my_stack.is_empty() << endl;
  cout << my_stack.pop() << endl;
  cout << my_stack.is_empty() << endl;
  my_stack.pop();

  return 0;
}

File: stacks/linked_list_stack.py
class Node:
  def __init__(self, d, n):
    self.data = d
    self.next = n

class Stack:
  def __init__(self):
    self.top_node = None

  def push(self, x):
    self.top_node = Node(x, self.top_node)

  def pop(self):
    if self.is_empty():
      print("Stack is empty")
    else:
      data = self.top_node.data
      self.top_node = self.top_node.next
      return data

  def peek(self):
    return self.top_node.data
  
  def is_empty(self):
    return self.top_node == None


# test
my_stack = Stack()
my_stack.push(4)
my_stack.push(-5)
print(my_stack.pop())
my_stack.push(99)
print(my_stack.peek())
print(my_stack.is_empty())